### ISSUE TITLE
Why defining bias term yourself?

### ISSUE THREAD

**Issue by SaoYan** [akanimax/pro_gan_pytorch#7]
Hi there, in CustomLayers.py, the bias term is defined by hand:
> self.bias = th.nn.Parameter(th.FloatTensor(c_out).fill_(0))

Why not just use bias=True/False in nn.Conv2d and nn.Linear?

**Comment by akanimax** [comment]
Hi @SaoYan,

For the Equalized learning rate as described in **`section 4.1`**, we need to keep separate copies of the weights and biases since the He constant is multiplied only with the weight of the `conv` and not the `biases`. If you notice that the bias for the pyTorch conv and convTranspose layer are turned off.
Hope this helps

Best regards,
@akanimax

**Comment by SaoYan** [comment]
Thanks for reply.

Actually I’m still confused about the implementation of equalized learning rate. 

In the paper, the equation is quite simple: w/c. My questions are:

1. Is self.scale the He constant? If so, why do weight/scale and then x.mul(scale)?

2. To my understanding the He constant in Kaiming’s paper is the so-called “fan_in”, so I cannot understand the compuation of self.scale in the code.

Thanks ~

**Comment by akanimax** [comment]
Yes you are right. The He constant's value is `sqrt(2 / fan_in)`. So, for calculating the scale, notice the trick that is being applied here:

-> Initialize the weights using "Kaiming Initializer". Thus the weights' parameters are drawn from __`N(0, c)`__ where `c = sqrt(2 / fan_in)`. Now for extracting the `c` from these initialized weights, I calculate the std as `self.scale = (th.mean(self.conv.weight.data ** 2)) ** 0.5`.

->  Then while computing the forward pass, I am multiplying the `x` with `c` instead of the weights due to the property of convolution. It results in the same output.

I agree that this implementation is complicated. Can you propose a simpler way of doing so? 
Can you open PR here?

Best regards,
@akanimax

**Comment by akanimax** [comment]
Also, btw, I am not sure if it is a notational mistake or whether I am missing something, but in the paper they describe setting `w' = w / c` while in their official code, they set `w' = w * c` (Refer [this](https://github.com/tkarras/progressive_growing_of_gans/blob/master/networks.py#L27) line of the official TF code).

Please let me know if you find about why this is.
Hope this helps

Best regards,
@akanimax

**Comment by SaoYan** [comment]
Thanks for the detailed reply.

Actually I have no better idea for now.

I don't think that line of code you pointed out is the full implementation of equalized learning rate. Considering that TF builds a static graph, that line only runs once during initialization. Though I'm not sure, I think the actual "run-time" parameter update is [in tfutil](https://github.com/tkarras/progressive_growing_of_gans/blob/master/tfutil.py#L246), where they re-implement the base class Optimizer in TF. It seems that there are lots of details in Optimizer, and I'll still trying to read it. I'd be happy to discuss this.

**Comment by akanimax** [comment]
I have actually gone through most of the Optimizer code as well. It doesn't have any additional code for the equalized learning rate. Indeed, the graphs are statically built, but notice how the call is made to `tf.get_variable()` inside the `get_weight()` function which is in turn called by the `conv2d()` function. The function `tf.get_variable()` initializes a new weight if the variable doesn't exist, else it returns the already present variable with the same name. basically, in the graph, everytime the conv operation is applied, they are multiplying the weights of the convolution by the `wscale` constant parameter.

Let me know if you find anything more.

**Comment by akanimax** [comment]
Closing this issue now...
Thank you for your contribution. I have made some major changes to the code now. Please do check it out.
